### ISSUE TITLE
fill-rule Could Use Some Clarification

### ISSUE THREAD

**Issue by patrickdark** [w3c/svgwg#344]
https://svgwg.org/svg2-draft/painting.html#FillRuleProperty

I just went through a thread on Stack Overflow (at https://stackoverflow.com/questions/1983256/how-can-i-cut-one-shape-inside-another) and realized that `fill-rule` applies to "shapes and text content elements" but the SVG 1.1/2 specifications don't provide any indication of how to use this property for anything other than `path` elements. The SVG 1.1 test suite only provides one test case and that likewise involves `path` elements: https://www.w3.org/Graphics/SVG/Test/20110816/svg/painting-fill-03-t.svg.

It seems that some clarification is needed on how to use this property with `ellipse` elements etc.

**Comment by svgeesus** [comment]
The sort of union and difference operators mentioned in the stack overflow question used to be a part of SVG vector effects; they were dropped form SVG2 unfortunately.
As the answers on stack overflow state, the component elements need to be paths (or text, in which case it is the geometry of the glyph outlines that are used). But only on a single element. SVG does not currently have a way to say "take this shape and then subtract from it this other shape".
fill-rule can be set on an ellipse, for example, but an ellipse is not self-intersecting so it has no effect.
Yes, the test suite should have tests added for fill-rule on shapes that can be self intersecting (like polyline) and for text elements.

**Comment by patrickdark** [comment]
@svgeesus Right. So, without regard to the `operation` attribute, the spec should do one of:

- (A) not allowing `fill-rule` attributes on elements where it has no effect;
- (B) explaining why it allows `fill-rule` attributes on elements where it has no effect; or
- (C) redefining `fill-rule` attributes so they they have an effect on elements where the attribute is currently permitted, yet does nothing.

Then do all of:
- (D) add an example and testcase or testcase component for every element on which the `fill-rule` attribute is allowed; and
- (E) point out why the two ellipse examples given in the SVG 2 (at https://svgwg.org/svg2-draft/images/painting/fillrule-nonzero.svg) spec use `path` elements instead of `ellipse` elements in the case of (A).

Regarding (B), I have to wonder if that's related to the old SVG DTD.

Regarding (C), in the Stack Overflow example, two answerers tried to use the `fill-rule` attribute on `g` elements, where it's prohibited in both SVG 1.1 and SVG 2. Allowing the attribute there would significantly increase its utility.

**Comment by AmeliaBR** [comment]
`fill-rule` is not merely an attribute; it is an inheritable style property and presentation attribute. It can be defined [on any SVG-namespaced element in SVG 2](https://svgwg.org/svg2-draft/styling.html#TermPresentationAttribute), and [most SVG-namespaced elements in SVG 1.1](https://www.w3.org/TR/SVG11/attindex.html#PresentationAttributes) regardless of whether or not it has an effect on that element.  In particular, it has _always_ been valid as an attribute on `<g>`; however, as @svgeesus explained, it does not apply at the group level, but by inheriting down to individual shape and text elements (just like the `fill` or `stroke` properties do).  It would therefore be unacceptable to change its behavior.

The "Applies to" section of the [`fill-rule` definition](https://svgwg.org/svg2-draft/painting.html#FillRuleProperty) _could_ be narrowed to only list the elements on which it would ever have a visible effect, since "Applies to" defines which elements actually use the style property. However, I personally see that as unnecessarily complicating matters.  The `fill-rule` applies to a `<circle>` in the exact same way that it applies to a `<path>` element describing the same circular shape.  It is OK that definition of `fill-rule` refers specifically to paths, because all the basic shapes are now defined in terms of ["equivalent paths"](https://svgwg.org/svg2-draft/paths.html#TermEquivalentPath).

I *do* agree that it would be very nice if SVG had a feature for creating compound shapes that are the intersection or union of multiple simple shapes.  But that is quite a more complicated matter than what `fill-rule` can do.

That said...

The application of `fill-rule` to _text_ does need to be clarified.

In the [Text chapter](https://svgwg.org/svg2-draft/text.html#TextRenderingOrder), we currently have an informative note that says "Since the fill-rule property does not apply to SVG text elements, the specific order of the subpaths within the equivalent path does not matter."  That contradicts the [`fill-rule` definition](https://svgwg.org/svg2-draft/painting.html#FillRuleProperty) that says it _does_ apply to text content elements.  We need to decide which one it is (probably by testing current SVG implementations to see what they are doing), and we definitely need some tests to match.

There should also be tests for `fill-rule` and `<polygon>` and `<polyline>`, and maybe (for completeness) the other basic shapes.

**Comment by patrickdark** [comment]
@AmeliaBR 
> fill-rule is not merely an attribute; it is an inheritable style property and presentation attribute. It can be defined on any SVG-namespaced element in SVG 2, and most SVG-namespaced elements in SVG 1.1 regardless of whether or not it has an effect on that element. In particular, it has always been valid as an attribute on <g>; however, as @svgeesus explained, it does not apply at the group level, but by inheriting down to individual shape and text elements (just like the fill or stroke properties do). It would therefore be unacceptable to change its behavior.

Good to know; I wasn’t aware of this detail. So `<g fill-rule="evenodd"/>` is valid even though the `fill-rule` rule CSS property has no effect on `g` elements.

>The "Applies to" section of the fill-rule definition could be narrowed to only list the elements on which it would ever have a visible effect, since "Applies to" defines which elements actually use the style property. However, I personally see that as unnecessarily complicating matters. The fill-rule applies to a <circle> in the exact same way that it applies to a <path> element describing the same circular shape. It is OK that definition of fill-rule refers specifically to paths, because all the basic shapes are now defined in terms of "equivalent paths".

I think the “Applies to” section should be narrowed. As it stands, it implies that code like `<circle … fill-rule="evenodd"/>` should do something even though this is guaranteed to do nothing.

At the very least, I think an explanation in the specification should be added to indicate why elements like `circle` are listed to avoid confusion (as indicated by item (B) in my previous post).

>I do agree that it would be very nice if SVG had a feature for creating compound shapes that are the intersection or union of multiple simple shapes. But that is quite a more complicated matter than what fill-rule can do.

In light of the `fill-rule` CSS property already applying to `g` elements, I’d guess it’s out of the question to redefine `fill-rule` to apply to `g` elements as a union of its child or descendant shapes because of backward-compatibility concerns.

> The application of fill-rule to text does need to be clarified.

> In the Text chapter, we currently have an informative note that says "Since the fill-rule property does not apply to SVG text elements, the specific order of the subpaths within the equivalent path does not matter." That contradicts the fill-rule definition that says it does apply to text content elements. We need to decide which one it is (probably by testing current SVG implementations to see what they are doing), and we definitely need some tests to match.

>There should also be tests for fill-rule and <polygon> and <polyline>, and maybe (for completeness) the other basic shapes.

I have to wonder whether this is leftover functionality from when SVG Fonts were still a thing.

**Comment by boggydigital** [comment]
Not blocking updated 2.0 CR publication - assigning 2.1 WD milestone

**Comment by RazrFalcon** [comment]
@AmeliaBR 

>We need to decide which one it is (probably by testing current SVG implementations to see what they are doing), and we definitely need some tests to match.

It's pretty hard to force, but looks like at least Inkscape has some problems with that.

```xml
<svg viewBox="0 0 200 200" xmlns="http://www.w3.org/2000/svg" font-family="Arial" font-size="64">
    <text x="40" y="100" fill-rule="evenodd">مرحبا</text>
    <rect id="frame" x="1" y="1" width="198" height="198" fill="none" stroke="black"/>
</svg>
```

![e-text-040](https://user-images.githubusercontent.com/725494/54433912-ae9eeb00-4735-11e9-8432-4296656c9072.png)

Another example:

```xml
<svg viewBox="0 0 200 200" xmlns="http://www.w3.org/2000/svg" font-family="Arial" font-size="64">
    <text x="40" y="100" fill-rule="evenodd" letter-spacing="-20">Text</text>
    <rect id="frame" x="1" y="1" width="198" height="198" fill="none" stroke="black"/>
</svg>
```

![e-text-041](https://user-images.githubusercontent.com/725494/54434258-819f0800-4736-11e9-9377-007d33db3dd0.png)

But not everyone supports `letter-spacing`.

A negative `dx` works fine in the Inkscape case, probably because it draws each character separately.

**Comment by Tavmjong** [comment]
This should be easy to "fix" in Inkscape. It doesn't make sense, in my opinion, to apply evenodd to text.
